text/0041-extensible-block-body.md

```diff
@@ -0,0 +1,357 @@
+- A `transfer_hashes` database would store the lists of transfer hashes keyed by `hash(transfer_hashes)`.
+- A `proposer` database would store the proposers keyed by `hash(proposer)`.
+
+The new way of hashing and new databases would start being used in a specific, chosen protocol version. The blocks in this protocol version and above would have their bodies hashed using the new way and be stored in the new databases. Older blocks would be kept in the old database.
+
+This is to ensure that we can still verify the integrity of storage after we transition to the new format. Changing the body hash of the old blocks would also change the hash of the block and hashes of all the later blocks, so it is not an option. So in order to be able to verify that the stored data still corresponds to the hashes from the blocks, we need to be able to calculate them the old way even after we start using the new way for new blocks.
+
+The code reading the blocks from storage would then have to be aware of the protocol version at which the transition happened, in order to read the block bodies from the correct databases regardless of whether it's a new or an old block.
+
+### The new block body hash and upgrades
+
+The change to the way of calculating the hash of the block header and body would be compatible with upgrades.
+
+If we keep the ability of calculating hashes the old way, alongside the new way, and the software is aware of the protocol version at which the transition happened, it can always calculate the hash using the way that is correct for the given block. Thus, the new versions of the software would be compatible with the old blocks. On the other hand, older versions of the software should never have to have anything to do with the new blocks - if a node progressed past the upgrade at which the new block body format has been introduced, even if it starts an old version of the software, it would quickly realize that it's outdated, shut itself down and prompt the launcher to start a newer version.
+
+### Hashing Primitives In Rust
+
+Here we present the hashing primitives used for constructing the new hashes.
+
+The first is a primitive for hashing pairs of hashes:
+
+```rust
+pub fn hash_pair(hash1: &Digest, hash2: &Digest) -> Digest {
+    let mut to_hash = [0; Digest::LENGTH * 2];
+    to_hash[..Digest::LENGTH].copy_from_slice(&(hash1.to_array())[..]);
+    to_hash[Digest::LENGTH..].copy_from_slice(&(hash2.to_array())[..]);
+    hash(&to_hash)
+}
+```
+
+The second is a primitive for hashing slices. This follows the scheme for hashing a `BlockBody` described in the previous section. It is intended for hashing vectors where we care about order.  This is suited to hashing extendable data structures. It uses a [right fold][1].
+
+```rust
+pub fn hash_slice_rfold(slice: &[Digest]) -> Digest {
+    slice
+        .iter()
+        .rfold(SENTINEL, |prev, next| hash_pair(next, &prev))
+}
```

Comment:
```suggestion
pub fn hash_slice_rfold(slice: &[Digest]) -> Digest {
    slice
        .iter()
        .rfold([2u8; HASH_SIZE], |prev, next| hash_pair(next, &prev))
}
```

As discussed, it makes it simpler to debug if there are separate sentinels for different hashing primitives.